compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -208,7 +208,18 @@ class ClassCodegen private constructor(
             entry is MultifileFacadeFileEntry -> KotlinClassHeader.Kind.MULTIFILE_CLASS
             else -> KotlinClassHeader.Kind.SYNTHETIC_CLASS
         }
-        writeKotlinMetadata(visitor, state, kind, extraFlags) {
+
+        // There are three kinds of classes which are regenerated during inlining.
+        // 1) Anonymous classes which are in the scope of an inline function.
+        // 2) SAM wrappers used in an inline function. These are identified by name, since they
+        //    can be reused in different functions and are thus generated in the enclosing top-level
+        //    class instead of inside of an inline function.
+        // 3) WhenMapping classes used from inline functions. These are collected in
+        //    `JvmBackendContext.publicAbiSymbols` in `MappedEnumWhenLowering`.
+        val isPublicAbi = irClass.symbol in context.publicAbiSymbols || type.isInlineSamWrapper ||
+                type.isAnonymousClass && irClass.isInPublicInlineScope
```

Comment:
`type.isAnonymousClass` could `irClass` be used instead of type here?

Reply:
Using `irClass.isAnonymousObject` ends up missing some cases here (e.g. in `testInnerMadness`). The logic needs to match the `AnonymousObjectTransformer` and this was the easiest way. However, we can look at `irClass` to identify inline sam wrappers and I've changed the code accordingly.

Reply:
@udalov we need to think which way is less fragile in long perspective. Current approach is looking OK